internal/secrets/encryptor.go

```diff
@@ -69,43 +69,94 @@ func (e encryptor) EncryptBytes(plaintext []byte) (ciphertext []byte, err error)
-		block, err := aes.NewCipher(key)
-		if err != nil {
-			return nil, err
-		}
+	gcm, err := cipher.NewGCM(block)
+	if err != nil {
+		return nil, err
+	}
 
-		gcm, err := cipher.NewGCM(block)
-		if err != nil {
-			return nil, err
-		}
+	if len(ciphertext) < gcm.NonceSize() {
+		return nil, errors.New("malformed ciphertext")
+	}
 
-		if len(ciphertext) < gcm.NonceSize() {
-			return nil, errors.New("malformed ciphertext")
-		}
+	return gcm.Open(nil, ciphertext[:gcm.NonceSize()], ciphertext[gcm.NonceSize():], nil)
+}
 
-		return gcm.Open(nil,
-			ciphertext[:gcm.NonceSize()],
-			ciphertext[gcm.NonceSize():],
-			nil,
-		)
+func newEncryptor(primaryKey, secondaryKey []byte) Encryptor {
+	return encryptor{
+		primaryKey:   primaryKey,
+		secondaryKey: secondaryKey,
 	}
-	if plaintext, err = decrypt(e.primaryKey, ciphertext); err == nil {
+}
+
+// ConfiguredToEncrypt returns the statue of our encryptor, whether or not
+// it has a key specified, and can thus encrypt.
+func (e encryptor) ConfiguredToEncrypt() bool {
+	return len(e.primaryKey) == validKeyLength
+}
+
+func (e encryptor) ConfiguredToRotate() bool {
+	return len(e.primaryKey) == validKeyLength && len(e.secondaryKey) == validKeyLength
+}
+
+// EncryptBytes encrypts the plaintext using the primaryKey of the encryptor. This
+// relies on the AES-GCM encryption defined in encrypt, within this package.
+func (e encryptor) EncryptBytes(plaintext []byte) (ciphertext []byte, err error) {
+	if len(e.primaryKey) < validKeyLength {
+		return nil, &EncryptionError{errors.New("primary key is unavailable")}
+	}
+
+	return gcmEncrypt(plaintext, e.primaryKey)
+}
+
+// DecryptBytes decrypts the plaintext using the primaryKey of the encryptor.
+// This relies on AES-GCM.
+func (e encryptor) DecryptBytes(ciphertext []byte) (plaintext []byte, err error) {
+	if len(e.primaryKey) < validKeyLength && len(e.secondaryKey) < validKeyLength {
+		return nil, &EncryptionError{errors.New("no valid keys available")}
+	}
+
+	if plaintext, err = gcmDecrypt(ciphertext, e.primaryKey); err == nil {
 		return plaintext, nil
 	}
-	if plaintext, err = decrypt(e.secondaryKey, ciphertext); err == nil {
+	if plaintext, err = gcmDecrypt(ciphertext, e.secondaryKey); err == nil {
 		return plaintext, nil
 	}
 	return nil, &EncryptionError{err}
+
+}
+
+func (e encryptor) EncryptWithKey(plaintext, key []byte) ([]byte, error) {
+	return gcmEncrypt(plaintext, key)
+}
+
+// RotateEncryption rotates the encryption on a ciphertext by
+// decrypting the byte array using the primaryKey, and then reencrypting
+// it using the secondaryKey.
+func (e encryptor) RotateEncryption(ciphertext []byte) ([]byte, error) {
+	if !e.ConfiguredToRotate() {
+		return nil, &EncryptionError{errors.New("key rotation not configured")}
+	}
+	// try previous key first
+	plaintext, err := gcmDecrypt(ciphertext, e.secondaryKey)
+	if err != nil {
+		_, err = gcmDecrypt(ciphertext, e.primaryKey)
+		if err != nil {
+			return ciphertext, err
```

Comment:
nit: Unless there is a specific reason to return it, which should also need a code comment to explain why.
```suggestion
			return nil, err
```

Reply:
@chayim I believe there was a specific reason we wanted to return this?

Reply:
No, the err check itself is valid.  Committed the suggestion and merging.